### DISCUSSION TITLE
Welcome to PyNCBIminer Discussions!

### DISCUSSION THREAD

**Discussion by Xiaoting-Xu** [Xiaoting-Xu/PyNCBIminer#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by hlstrived** [comment]
Dear Xiaoting-Xu,

I hope this message finds you well. I am a user of your software, PyNCBIminer, and I would like to start by expressing my gratitude for developing such a valuable tool, which plays a significant role in my research. However, I have encountered some issues during its use and am reaching out for your assistance.

Specifically, when running PyNCBIminer for data analysis, the software stops responding after some time. The interface remains unchanged, displaying neither error messages nor output results; it simply freezes. This issue occurs every time I use the software, preventing me from completing the necessary data analyses.
![1](https://github.com/Xiaoting-Xu/PyNCBIminer/assets/94972506/e442b0ce-3a5c-4db8-bd06-4e0b71fe3798)
![2](https://github.com/Xiaoting-Xu/PyNCBIminer/assets/94972506/02d0e9c1-0a3f-485c-ace2-5cc10933a2d3)
![3](https://github.com/Xiaoting-Xu/PyNCBIminer/assets/94972506/ae425898-8122-498f-966f-58d100363ea5)
![4](https://github.com/Xiaoting-Xu/PyNCBIminer/assets/94972506/533fc6da-fd37-4ce6-bafb-209e9d891605)

To ensure that this problem was not caused by my operational or environmental setup errors, I have taken the following steps:

1. Checked and confirmed that my internet connection is stable;
2. Operated according to the software documentation to ensure correct use；
3. Attempted to run the software on different computers, encountering the same issue.

I sincerely hope to receive your assistance in resolving this matter. Should you need any more details about the specific circumstances of the issue or my system environment, please do not hesitate to ask. Your support is crucial for the continuation of my research work.

Thank you again for your contribution to the scientific community, and I look forward to your reply.

Best wishes,
Ling

**Comment by saphira-git** [comment]
Hi there,

thanks for creating this fabulous software.
I have a question about how the target taxon groups are treated. I have been inputting a list of species names in this format
Genus1 species1
Genus1 species2
Genus2 species1

However when comparing the sequences retrieved to the species list I searched for, it appears some accessions were pulled which weren't species I was searching for. I have checked these individually and it's not a case of a synonym being attached to the accession. How does pyNCBIminer search for the target taxons, does it accept partial matches such that it might pull a genus match without the species matching? It's weird because from a manual search, if it was doing this, it's not getting all of them (gene accessions for a genus match available on genbank), so I'm just not sure how it's working on the back-end.

Thanks
Saphira

**Comment by Xiaoting-Xu** [comment]
Dear Saphira,

Thanks for this feedback. Please enter your target taxon groups using only genus names or species names. PyNCBIminer treats one line as a single taxonomic name, so I recommend searching and downloading by genus first, then selecting your target species from each genus. This approach avoids numerous BLAST searches, which can be slow.

Best wishes,
Xiaoting

-----Original Messages-----
From:"Saphira Bloom-Quinn" ***@***.***>
Send time:Tuesday, 02/10/2026 11:48:18
To: Xiaoting-Xu/PyNCBIminer ***@***.***>
Cc: Xiaoting-Xu ***@***.***>, Author ***@***.***>
Subject: Re: [Xiaoting-Xu/PyNCBIminer] Welcome to PyNCBIminer Discussions! (Discussion #1)






Hi there,

thanks for creating this fabulous software.
I have a question about how the target taxon groups are treated. I have been inputting a list of species names in this format
Genus1 species1
Genus1 species2
Genus2 species1

However when comparing the sequences retrieved to the species list I searched for, it appears some accessions were pulled which weren't species I was searching for. I have checked these individually and it's not a case of a synonym being attached to the accession. How does pyNCBIminer search for the target taxons, does it accept partial matches such that it might pull a genus match without the species matching? It's weird because from a manual search, if it was doing this, it's not getting all of them (gene accessions for a genus match available on genbank), so I'm just not sure how it's working on the back-end.

Thanks
Saphira

**Comment by saphira-git** [comment]
Hi Xiaoting,
Thanks for your quick reply. You said pyNCBIminer treats each line as a single taxonomic name, but that still leaves me confused why I got sequences that were not in the list of taxonomic names I was searching for. For example, I searched for Dawsonia polytrichoides and Dawsonia superba, yet my fasta file also includes Dawsonia becarri, which is not a synonym of the other two names. Why would it have downloaded this sequence? 
one thing I really like about pyNCBIminer is that it is very automated. I am searching for something like 2000 species names as the study I am doing is for a whole phylum of organisms. Your suggestion is to search by genus and then select the target species after, to reduce the time required for numerous BLAST searches. Do you have any suggestion for automating this? I am trying to avoid manually having to delete lots of individual sequences from very large fasta files.
Best,
Saphira

**Reply by saphira-git** [reply]
I would like to provide an update to this comment as I believe I have resolved several aspects for it. My understanding as to why pyNCBIminer was extracting not-target species is because, as I understand it, the target groups are only used for the initial BLAST search. The consecutive rounds of BLAST focus on retrieving more sequences based only on the actual sequences retrieved from the first round. The later BLAST rounds no longer incorporate the taxonomic groups as in the first round; so, if another species has close sequence similarity to a target taxon, the sequence may be retrieved through this process. (Xiaoting, can you confirm this?). 

In terms of automating the filtering of non-target taxonomic groups from the final pyNCBIminer files, I found this was very easy to do using seqkit with the command seqkit grep -r -f target_species_list.txt final.fasta > extracted_species.fasta

**Reply by wyx619** [reply]
Hi Saphira,

Thank you for the detailed follow-up, and your understanding is surely correct!

You have correctly identified the root cause. In pyNCBIminer, the iterated BLAST process works as follows:

1. Round 1 : The search uses your target taxonomic names (e.g., Dawsonia polytrichoides[Organism] OR Dawsonia superba[Organism] ) to retrieve initial sequences.
2. Rounds 2+ : Subsequent BLAST rounds use sequence similarity (based on the sequences retrieved from previous rounds) rather than taxonomic filtering. This means if a sequence from Dawsonia becarri shares high similarity with your query sequences, it can be retrieved—even though Dawsonia becarri was not in your original target list.
This design allows the tool to discover related sequences that might not be easily found through taxonomic queries alone. However, I understand this behavior may be undesirable if your goal is strictly to retrieve sequences only from specific taxa.

I hope this clarifies. Additionally, I'm glad to hear that seqkit worked well for your post-filtering needs!

You raise an important point about the design philosophy of pyNCBIminer. Our primary goal is to maximize sequence coverage and diversity during the download process. The iterated BLAST approach is intentionally designed to be broad, capturing related sequences that might be missed by strict taxonomic queries alone.

Given this philosophy, performing post-filtering on the final FASTA file using tools like seqkit is indeed a more flexible and effective approach. It gives you full control over your filtering criteria without limiting the initial discovery process.

We believe this strategy—broad download followed by user-defined filtering—provides the best balance between comprehensive data acquisition and precise downstream analysis.

Thank you for sharing your workflow, and please don't hesitate to reach out if you have other questions or suggestions!

Best,
Yuxuan,
Collaborator of PyNCBIminer
